File: src/game/creature/TurretZombie.java
package game.creature;
import game.TextureManager;
import game.Vector;
import game.level.Level;

import java.util.ArrayList;
import java.util.List;


public class TurretZombie extends Zombie {
	public Turret head = new Turret();
	float run = -1;
	Vector runDir;
	float lastRun = 0;
	
	boolean spawnReq = false;
	Vector spawnPos;
	
	List<Creature> spawns = new ArrayList<>();
	
	
	public TurretZombie() {
		drawRadius *= 2;
		hitRadius = drawRadius - 0.1f;
		head.drawRadius = 0.45f * drawRadius;
		head.hitRadius = hitRadius;
		head.bulletSpeed = 18;
		texture = TextureManager.ZOMBIEBODY;
		walkSpeed = 60;
		mass = 100;
		health = 10;
	}
	
	@Override
	public void update(float dt, Level level) {
		head.pos = pos;
		head.invincible = invincible;
		if (level.tileGrid.hitsWall(pos, level.player.pos) || run > 0)
			head.angle = angle;

		head.update(dt, level);

		
		if (run > 0) {
			lastRun = (float) Math.random() * 5 + 1;
			
			Vector facing = new Vector(angle);
			int dir;
			if (runDir.dot(facing.perpCCW()) > 0) dir = 1;
			else dir = -1;
			
			angle += 3*dir*dt;
			if (run < 1)
				vel = vel.add(new Vector(angle).scale(walkSpeed*3*dt));
			
			run -= dt;
			
			creatureUpdate(dt, level);
		} else {
			super.update(dt, level);
		}
		
		if (spawnReq && spawnPos.distance(pos) > hitRadius) {
			spawnReq = false;
			Creature c;
			if (health <= 3) {
				c = new Turret();
			} else {
				c = new Zombie();
			}
			c.pos = spawnPos;
			c.angle = (float) (Math.random() * 2*Math.PI);
			level.spawnReqs.add(c);
			spawns.add(c);
		}
		
		lastRun -= dt;
		if (lastRun < 0 && detectedPlayer) {
			run = 1.5f;
			runDir = level.player.pos.sub(pos).direction();
		}
	}
	
	@Override
	public void takeHit(int damage, Vector dir) {
		int oldHealth = health;
		super.takeHit(damage, dir);
		if (oldHealth == health) return;
		
		if (health == 0) {
			for (Creature c : spawns) {
				c.health = 0;
			}
			return;
		}
		
		run = 1.5f;
		runDir = dir.negate().direction();
		
		spawnReq = true;
		spawnPos = pos;
		
		invincible = run;
	}
	
	@Override
	public void render() {
		head.pos = pos;
		head.invincible = invincible;
		
		super.render();
		head.render();
	}
}

File: ts/game/creature/TurretZombie.ts
import * as PIXI from 'pixi.js';
import { Vector } from '../Vector.js';
import { Level } from '../level/Level.js';
import { Creature } from './Creature.js';
import { Zombie } from './Zombie.js';
import { Turret } from './Turret.js';
import { TextureManager } from '../TextureManager.js';

export class TurretZombie extends Zombie {
  head = new Turret();
  run = -1;
  runDir = new Vector(0, 0);
  lastRun = 0;
  spawnReq = false;
  spawnPos = new Vector(0, 0);
  spawns: Creature[] = [];

  constructor() {
    super();
    this.drawRadius *= 2;
    this.hitRadius = this.drawRadius - 0.1;
    this.head.drawRadius = 0.45 * this.drawRadius;
    this.head.hitRadius = this.hitRadius;
    this.head.bulletSpeed = 18;
    this.texture = TextureManager.ZOMBIEBODY;
    this.walkSpeed = 60;
    this.mass = 100;
    this.health = 10;
  }

  update(dt: number, level: Level) {
    this.head.pos = this.pos;
    this.head.invincible = this.invincible;
    if (level.tileGrid.hitsWall(this.pos, level.player.pos) || this.run > 0) {
      this.head.angle = this.angle;
    }
    this.head.update(dt, level);

    if (this.run > 0) {
      this.lastRun = Math.random() * 5 + 1;
      const facing = new Vector(this.angle);
      const dir = this.runDir.dot(facing.perpCCW()) > 0 ? 1 : -1;
      this.angle += 3 * dir * dt;
      if (this.run < 1) this.vel = this.vel.add(new Vector(this.angle).scale(this.walkSpeed * 3 * dt));
      this.run -= dt;
      super.update(dt, level);
    } else {
      super.update(dt, level);
    }

    if (this.spawnReq && this.spawnPos.distance(this.pos) > this.hitRadius) {
      this.spawnReq = false;
      let c: Creature;
      if (this.health <= 3) c = new Turret();
      else c = new Zombie();
      c.pos = this.spawnPos;
      c.angle = Math.random() * 2 * Math.PI;
      level.spawnReqs.push(c);
      this.spawns.push(c);
    }

    this.lastRun -= dt;
    if (this.lastRun < 0 && this.detectedPlayer) {
      this.run = 1.5;
      this.runDir = level.player.pos.sub(this.pos).direction();
    }
  }

  takeHit(damage: number, dir: Vector) {
    const oldHealth = this.health;
    super.takeHit(damage, dir);
    if (oldHealth === this.health) return;
    if (this.health === 0) {
      for (const c of this.spawns) {
        c.health = 0;
      }
      return;
    }
    this.run = 1.5;
    this.runDir = dir.negate().direction();
    this.spawnReq = true;
    this.spawnPos = this.pos;
    this.invincible = this.run;
  }

  render(container: PIXI.Container) {
    this.head.pos = this.pos;
    this.head.invincible = this.invincible;
    super.render(container);
    this.head.render(container);
  }
}
